bin/main.rb

```diff
@@ -1 +1,150 @@
+    puts 'Choose numbers from 1 to 9 to select desired cell'
+    puts 'No duplicate numbers are allowed'
+    puts '--------------------------------'
+    puts 'Player 1: What is your name?:'
+    @player_one = gets.chomp
+    puts 'Player 2: What is your name?:'
+    @player_two = gets.chomp
+    puts '************************************************************'
+    puts "Player one X is #{player_one} and Player two O #{player_two}"
+    puts '************************************************************'
+    puts 'Game Start!'
+  end
+
+  def draw_player_turn(player)
+    puts '************************************************************'
+    puts "*                  #{player} its your turn!                *"
+    puts '************************************************************'
+  end
+
+  def draw_available_moves(array)
+    puts '************************************************************'
+    puts "*       This are the available moves  #{array}             *"
+    puts '************************************************************'
+  end
+
+  def draw_invalid_move
+    puts '************************************************************'
+    puts '*                This move is not valid!                   *'
+    puts '************************************************************'
+  end
+
+  def draw_win
+    puts '************************************************************'
+    puts '*                This is a winning move!                   *'
+    puts '************************************************************'
+  end
+
+  def draw_itsa_draw
+    puts '************************************************************'
+    puts '*                    This is a draw!                       *'
+    puts '************************************************************'
+  end
+
+  def draw?
+    true if @total_moves == 0
+  end
+
+  def invalid_move?
+    false
+  end
+
+  def win?
+    false
+  end
+end
+
+#########
+game_on = true
+# initial instructions for player
+game = DisplayInterface.new
+game.draw_welcome
+player_one = game.player_one
+player_two = game.player_two
+while game_on
+  # loop for each move
+  if game.get_current_player == player_one 
+    if game.draw?
+      game.draw_itsa_draw
+      game_on = false
+    elsif game.invalid_move?
+    game.draw_invalid_move
+    elsif game.win?
+      game.draw_win
+    end
+    if game_on
+      game.draw_player_turn(player_one)
+      position = game.get_position.to_i
+      game.take_place(position, 'X') 
+      game.draw_board 
+    end
+  else 
+    if game.draw?
+      game.draw_itsa_draw
+      game_on = false
+    elsif game.invalid_move?
+    game.draw_invalid_move
+    elsif game.win?
+      game.draw_win
+    end
+    if game_on
+      game.draw_player_turn(player_two)
+      position = game.get_position.to_i
+      game.take_place(position, 'O') 
+      game.draw_board
+    end
+  end
+  game.draw_line
+end
```

Comment:
- Since you used a class for user interface, it's better to include the loop inside the userinterface class as a method and call it after initializing its object.
- The logic of the game loop should give the game the following features:
  - Ask the player for a move, if it's valid, put the symbol on the board, else it should ask again for a valid move. Use conditionals to achieve that.
  - For each valid move, check for the win or the draw for each player and display a message for the concerned player.
   - If the move is valid, it should do the same with the other player (i.e it goes on like that)

I noticed these flaws in the game loop after and changed to true each game condition methods :red_circle: separately.

Reply:
- We put the while loop inside a method and called it after initializing the game object.

- However the logic that you are asking us to implement is a requirement of the game logic and we already implemented that in the game logic